src/python/pants/core/util_rules/lockfile_metadata.py

```diff
@@ -72,18 +74,24 @@ class LockfileMetadata:
     To construct an instance of the most recent concrete subclass, call `LockfileMetadata.new()`.
     """
 
-    _LockfileMetadataSubclass = TypeVar("_LockfileMetadataSubclass", bound="LockfileMetadata")
-
+    # N.B. This __must not__ change between multiple versions of a lockfile metadata for a
+    # particular scope. The delimeter for version 1 must be the delimeter for version N.
+    delimeter: ClassVar[str] = "#"
```

Comment:
We can remove the default and force subclasses to provide this. Dealer's choice.